core/include/traccc/fitting/kalman_filter/two_filters_smoother.hpp

```diff
@@ -0,0 +1,173 @@
+template <typename algebra_t>
+struct two_filters_smoother {
+
+    // Type declarations
+    using size_type = detray::dsize_type<algebra_t>;
+    template <size_type ROWS, size_type COLS>
+    using matrix_type = detray::dmatrix<algebra_t, ROWS, COLS>;
+    using bound_vector_type = detray::bound_vector<algebra_t>;
+    using bound_matrix_type = detray::bound_matrix<algebra_t>;
+
+    /// Two-filters smoother operation
+    ///
+    /// @param mask_group mask group that contains the mask of surface
+    /// @param index mask index of surface
+    /// @param trk_state track state of the surface
+    /// @param bound_params bound parameter
+    ///
+    /// @return true if the update succeeds
+    template <typename mask_group_t, typename index_t>
+    TRACCC_HOST_DEVICE inline bool operator()(
+        const mask_group_t& /*mask_group*/, const index_t& /*index*/,
+        track_state<algebra_t>& trk_state,
+        bound_track_parameters& bound_params) const {
+
+        using shape_type = typename mask_group_t::value_type::shape;
+
+        const auto D = trk_state.get_measurement().meas_dim;
+        assert(D == 1u || D == 2u);
+        if (D == 1u) {
+            return smoothe<1u, shape_type>(trk_state, bound_params);
+        } else if (D == 2u) {
+            return smoothe<2u, shape_type>(trk_state, bound_params);
+        }
+
+        return false;
+    }
+
+    // Reference: The Optimun Linear Smoother as a Combination of Two Optimum
+    // Linear Filters
+    template <size_type D, typename shape_t>
+    TRACCC_HOST_DEVICE inline bool smoothe(
+        track_state<algebra_t>& trk_state,
+        bound_track_parameters& bound_params) const {
+
+        assert(trk_state.filtered().surface_link() ==
+               bound_params.surface_link());
+
+        static_assert(((D == 1u) || (D == 2u)),
+                      "The measurement dimension should be 1 or 2");
+
+        const auto meas = trk_state.get_measurement();
+
+        matrix_type<D, e_bound_size> H = meas.subs.template projector<D>();
+
+        // Measurement data on surface
+        const matrix_type<D, 1>& meas_local =
+            trk_state.template measurement_local<D>();
+
+        // Predicted vector of bound track parameters
+        const matrix_type<e_bound_size, 1> predicted_vec =
+            bound_params.vector();
+
+        // Predicted covaraince of bound track parameters
+        const matrix_type<e_bound_size, e_bound_size> predicted_cov =
+            bound_params.covariance();
+
+        const matrix_type<e_bound_size, e_bound_size> predicted_cov_inv =
+            matrix::inverse(predicted_cov);
+        const matrix_type<e_bound_size, e_bound_size> filtered_cov_inv =
+            matrix::inverse(trk_state.filtered().covariance());
+
+        // Eq (3.38) of "Pattern Recognition, Tracking and Vertex
+        // Reconstruction in Particle Detectors"
+        const matrix_type<e_bound_size, e_bound_size> smoothed_cov_inv =
```

Comment:
Is this not a 
```suggestion
        const bound_matrix_type smoothed_cov_inv =
```
?
Same in other places and also for the ```bound_vector_type```

Reply:
To be honest, I prefer using `matrix_type<int, int>`, which is more informative to readers...

Reply:
Sure enough! Having just gone through the algebra plugins templating cleanup, I slightly prefer centrally defined type aliases that I don't have to chase through the code, but that is a matter of taste I guess :)